File: src/main/java/org/anhonesteffort/btc/strategy/MetaStrategy.java
/*
 * Copyright (C) 2016 An Honest Effort LLC.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.anhonesteffort.btc.strategy;

import org.anhonesteffort.btc.book.Order;
import org.anhonesteffort.btc.http.request.model.PostOrderRequest;
import org.anhonesteffort.btc.state.GdaxState;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Optional;

public class MetaStrategy extends Strategy<Void> {

  private static final Logger log = LoggerFactory.getLogger(MetaStrategy.class);

  private final StrategyFactory        strategies;
  private final BidIdentifyingStrategy bidIdStrategy;
  private final AskIdentifyingStrategy askIdStrategy;

  private OrderOpeningStrategy   openStrategy;
  private OrderMatchingStrategy  matchStrategy;
  private OrderCancelingStrategy cancelStrategy;

  private State state;
  private Order bidPosition;
  private Order askPosition;

  private enum State {
    IDENTIFY_BID, OPENING_BID, MATCHING_BID,
    IDENTIFY_ASK, OPENING_ASK, MATCHING_ASK,
    COMPLETE, ABORT
  }

  public MetaStrategy(StrategyFactory strategies) {
    this.strategies = strategies;
    bidIdStrategy   = strategies.newBidIdentifying();
    askIdStrategy   = strategies.newAskIdentifying();
    state           = State.COMPLETE;

    addChildren(bidIdStrategy, askIdStrategy);
  }

  @Override
  protected Void advanceStrategy(GdaxState bookState, long nanoseconds) {
    switch (state) {
      case COMPLETE:
        log.info("awaiting buy opportunity");
        askIdStrategy.setContext(Optional.empty(), Optional.empty());
        bidPosition = null;
        askPosition = null;
        state       = State.IDENTIFY_BID;
        break;

      case IDENTIFY_BID:
        Optional<PostOrderRequest> postBid = bidIdStrategy.getResult();
        if (postBid.isPresent()) {
          log.info("opening bid for " + postBid.get().getSize() + " at " + postBid.get().getPrice());
          openStrategy = strategies.newOrderOpening(postBid.get());
          addChildren(openStrategy);
          state = State.OPENING_BID;
        }
        break;

      case OPENING_BID:
        Optional<Order> openedBid = openStrategy.getResult();
        if (openStrategy.isAborted()) {
          log.info("bid rejected due to post-only flag");
          removeChildren(openStrategy);
          state = State.IDENTIFY_BID;
        } else if (openedBid.isPresent()) {
          bidPosition = openedBid.get();
          log.info("bid opened with id " + bidPosition.getOrderId() + ", waiting to match");
          removeChildren(openStrategy);
          matchStrategy = strategies.newOrderMatching(Order.Side.BID, bidPosition.getOrderId());
          addChildren(matchStrategy);
          state = State.MATCHING_BID;
        }
        break;

      case MATCHING_BID:
        if (matchStrategy.isAborted()) {
          log.info("aborting bid due to match timeout");
          removeChildren(matchStrategy);
          cancelStrategy = strategies.newOrderCanceling(bidPosition.getOrderId());
          addChildren(cancelStrategy);
          state = State.ABORT;
        } else if (matchStrategy.getResult()) {
          log.info("bid matched with ask, awaiting sell opportunity");
          removeChildren(matchStrategy);
          askIdStrategy.setContext(Optional.of(bidPosition), Optional.empty());
          state = State.IDENTIFY_ASK;
        }
        break;

      case IDENTIFY_ASK:
        Optional<PostOrderRequest> postAsk = askIdStrategy.getResult();
        if (postAsk.isPresent()) {
          log.info("opening ask for " + postAsk.get().getSize() + " at " + postAsk.get().getPrice());
          openStrategy = strategies.newOrderOpening(postAsk.get());
          addChildren(openStrategy);
          state = State.OPENING_ASK;
        }
        break;

      case OPENING_ASK:
        Optional<Order> openedAsk = openStrategy.getResult();
        if (openStrategy.isAborted()) {
          log.info("ask rejected due to post-only flag");
          removeChildren(openStrategy);
          askIdStrategy.setContext(Optional.of(bidPosition), Optional.empty());
          state = State.IDENTIFY_ASK;
        } else if (openedAsk.isPresent()) {
          askPosition = openedAsk.get();
          log.info("ask opened with id " + askPosition.getOrderId() + ", waiting to match");
          removeChildren(openStrategy);
          matchStrategy = strategies.newOrderMatching(Order.Side.ASK, askPosition.getOrderId());
          addChildren(matchStrategy);
          state = State.MATCHING_ASK;
        }
        break;

      case MATCHING_ASK:
        if (matchStrategy.isAborted()) {
          log.info("aborting ask due to match timeout");
          removeChildren(matchStrategy);
          cancelStrategy = strategies.newOrderCanceling(askPosition.getOrderId());
          addChildren(cancelStrategy);
          state = State.ABORT;
        } else if (matchStrategy.getResult()) {
          log.info("ask matched with bid, strategy complete");
          removeChildren(matchStrategy);
          state = State.COMPLETE;
        }
        break;

      case ABORT:
        if (cancelStrategy.getResult() && askPosition == null) {
          log.info("bid " + bidPosition.getOrderId() + " canceled");
          removeChildren(cancelStrategy);
          state = State.IDENTIFY_BID;
        } else if (cancelStrategy.getResult()) {
          log.info("ask " + askPosition.getOrderId() + " canceled");
          removeChildren(cancelStrategy);
          askIdStrategy.setContext(Optional.of(bidPosition), Optional.of(askPosition));
          state = State.IDENTIFY_ASK;
        }
        break;
    }

    return null;
  }

}

File: jvm/src/main/scala/org/anhonesteffort/trading/strategy/MetaStrategy.scala
package org.anhonesteffort.trading.strategy
import org.anhonesteffort.trading.book.Orders.{Order, Side}
import org.anhonesteffort.trading.state.GdaxState
import org.slf4j.{Logger, LoggerFactory}

class MetaStrategy(strategies: StrategyFactory) extends Strategy[Unit] {

  private val log: Logger = LoggerFactory.getLogger(classOf[MetaStrategy])

  sealed trait State
  object State {
    case object IDENTIFY_BID extends State
    case object OPENING_BID  extends State
    case object MATCHING_BID extends State
    case object IDENTIFY_ASK extends State
    case object OPENING_ASK  extends State
    case object MATCHING_ASK extends State
    case object COMPLETE     extends State
    case object ABORT        extends State
  }

  private val bidIdStrategy : BidIdentifyingStrategy = strategies.newBidIdentifying
  private val askIdStrategy : AskIdentifyingStrategy = strategies.newAskIdentifying

  private var openStrategy   : OrderOpeningStrategy   = _
  private var matchStrategy  : OrderMatchingStrategy  = _
  private var cancelStrategy : OrderCancelingStrategy = _

  private var state       : State = State.COMPLETE
  private var bidPosition : Order = _
  private var askPosition : Order = _

  addChildren(Seq(bidIdStrategy, askIdStrategy))

  override protected def advanceStrategy(bookState: GdaxState, ns: Long): Unit = {
    state match {
      case State.COMPLETE =>
        log.info("awaiting buy opportunity")
        askIdStrategy.setContext(None, None)
        bidPosition = null
        askPosition = null
        state       = State.IDENTIFY_BID

      case State.IDENTIFY_BID =>
        bidIdStrategy.getResult match {
          case None      => Unit
          case Some(bid) =>
            log.info(s"opening bid for ${bid.getSize} at ${bid.getPrice}")
            openStrategy = strategies.newOrderOpening(bid)
            addChild(openStrategy)
            state = State.OPENING_BID
        }

      case State.OPENING_BID =>
        val bid = openStrategy.getResult
        if (openStrategy.isAborted) {
          log.info("bid rejected due to post-only flag")
          removeChild(openStrategy)
          state = State.IDENTIFY_BID
        } else if (bid.isDefined) {
          bidPosition = bid.get
          log.info(s"bid opened with id ${bidPosition.getOrderId}, waiting to match")
          removeChild(openStrategy)
          matchStrategy = strategies.newOrderMatching(Side.BID, bidPosition.getOrderId)
          addChild(matchStrategy)
          state = State.MATCHING_BID
        }

      case State.MATCHING_BID =>
        if (matchStrategy.isAborted) {
          log.info("aborting bid due to match timeout")
          removeChild(matchStrategy)
          cancelStrategy = strategies.newOrderCanceling(bidPosition.getOrderId)
          addChild(cancelStrategy)
          state = State.ABORT
        } else if (matchStrategy.getResult) {
          log.info("bid matched with ask, awaiting sell opportunity")
          removeChild(matchStrategy)
          askIdStrategy.setContext(Some(bidPosition), None)
          state = State.IDENTIFY_ASK
        }

      case State.IDENTIFY_ASK =>
        askIdStrategy.getResult match {
          case None      => Unit
          case Some(ask) =>
            log.info(s"opening ask for ${ask.getSize} at ${ask.getPrice}")
            openStrategy = strategies.newOrderOpening(ask)
            addChild(openStrategy)
            state = State.OPENING_ASK
        }

      case State.OPENING_ASK =>
        val ask = openStrategy.getResult
        if (openStrategy.isAborted) {
          log.info("ask rejected due to post-only flag")
          removeChild(openStrategy)
          askIdStrategy.setContext(Some(bidPosition), None)
          state = State.IDENTIFY_ASK
        } else if (ask.isDefined) {
          askPosition = ask.get
          log.info(s"ask opened with id ${askPosition.getOrderId}, waiting to match")
          removeChild(openStrategy)
          matchStrategy = strategies.newOrderMatching(Side.ASK, askPosition.getOrderId)
          addChild(matchStrategy)
          state = State.MATCHING_ASK
        }

      case State.MATCHING_ASK =>
        if (matchStrategy.isAborted) {
          log.info("aborting ask due to match timeout")
          removeChild(matchStrategy)
          cancelStrategy = strategies.newOrderCanceling(askPosition.getOrderId)
          addChild(cancelStrategy)
          state = State.ABORT
        } else if (matchStrategy.getResult) {
          log.info("ask matched with bid, strategy complete")
          removeChild(matchStrategy)
          state = State.COMPLETE
        }

      case State.ABORT =>
        if (cancelStrategy.getResult && askPosition == null) {
          log.info("bid " + bidPosition.getOrderId + " canceled")
          removeChild(cancelStrategy)
          state = State.IDENTIFY_BID
        } else if (cancelStrategy.getResult) {
          log.info("ask " + askPosition.getOrderId + " canceled")
          removeChild(cancelStrategy)
          askIdStrategy.setContext(Some(bidPosition), Some(askPosition))
          state = State.IDENTIFY_ASK
        }
    }
  }

}
